Parallel formatting

nikitamikhaylov opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- parallel parsing was rewritten to processors
- parallel formatting was implemented

nikitamikhaylov committed:
save changes

nikitamikhaylov committed:
fix JSONEachRowArray

nikitamikhaylov committed:
save

nikitamikhaylov committed:
first try

nikitamikhaylov committed:
remove CSV restriction

nikitamikhaylov committed:
fix build

nikitamikhaylov committed:
fix build

nikitamikhaylov committed:
remove stacktrace output

nikitamikhaylov committed:
fix build final

nikitamikhaylov committed:
save changes

nikitamikhaylov committed:
save

nikitamikhaylov committed:
save

nikitamikhaylov commented:
```sql
jakalletti.N1 :) create table tsv(a int, b int default 7) engine File(TSV);

CREATE TABLE tsv
(
    `a` int,
    `b` int DEFAULT 7
)
ENGINE = File(TSV)

Ok.

0 rows in set. Elapsed: 0.010 sec. 

jakalletti.N1 :) insert into tsv(a) select number from numbers(10000000);

INSERT INTO tsv (a) SELECT number
FROM numbers(10000000)

Ok.

0 rows in set. Elapsed: 2.963 sec. Processed 10.02 million rows, 80.18 MB (3.38 million rows/s., 27.06 MB/s.)
```
8 threads by default. After enabling the setting, drop the table and PageCache:
```sql
jakalletti.N1 :) set output_format_parallel_formatting = 1

SET output_format_parallel_formatting = 1

Ok.

0 rows in set. Elapsed: 0.002 sec. 

jakalletti.N1 :) insert into tsv(a) select number from numbers(10000000);

INSERT INTO tsv (a) SELECT number
FROM numbers(10000000)

Ok.

0 rows in set. Elapsed: 0.482 sec. Processed 10.02 million rows, 80.18 MB (20.78 million rows/s., 166.23 MB/s.) 
```

nikitamikhaylov committed:
it works

nikitamikhaylov committed:
better

nikitamikhaylov committed:
fix build

tavplubix commented on src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h:
```diff
@@ -0,0 +1,265 @@
+namespace DB
+{
+
+const size_t min_chunk_bytes_for_parallel_formatting = 1024;
```
It seems to be never used

nikitamikhaylov replied:
Sure :)

tavplubix commented on src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h:
```diff
@@ -0,0 +1,265 @@
+{
+
+const size_t min_chunk_bytes_for_parallel_formatting = 1024;
+const size_t max_threads_for_parallel_formatting = 6;
```
Why this restriction is required?

nikitamikhaylov replied:
I've got rid about it.

tavplubix commented on src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h:
```diff
@@ -0,0 +1,265 @@
+        , pool(max_threads_for_parallel_formatting)
+
+    {
+        processing_units.resize(max_threads_for_parallel_formatting + 2);
```
Please add comment

tavplubix commented on src/Processors/Formats/IOutputFormat.h:
```diff
@@ -39,12 +39,15 @@ class IOutputFormat : public IProcessor
 
     RowsBeforeLimitCounterPtr rows_before_limit_counter;
 
+    friend class ParallelFormattingOutputFormat;
```
Seems like it's not required

nikitamikhaylov replied:
We call `consume` inside and it is a protected member of `IOutputFormat`

tavplubix commented on src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h:
```diff
@@ -0,0 +1,265 @@
+
+            auto formatter = internal_formatter_creator(out_buffer);
+
+            formatter->consume(std::move(unit.chunk));
```
It may not work correctly for formats with non-trivial `doWritePrefix()` and `doWriteSuffix()`. Maybe disable parallel formatting for such formats?

nikitamikhaylov replied:
For now I've added `doWritePrefix()` and `doWriteSuffix()` methods, but it won't still work correctly for formats with stateful parsers (such as JSON). But it has to work with `CSVWithNamesAndTypes`, etc.

nikitamikhaylov committed:
delete and fix strange code

nikitamikhaylov committed:
better

nikitamikhaylov committed:
better

nikitamikhaylov committed:
add extremes and totals

nikitamikhaylov committed:
disable PrettySpaceMonoBlock + writePrefix

nikitamikhaylov committed:
better

nikitamikhaylov committed:
Update src/Formats/JSONEachRowUtils.cpp

Co-authored-by: tavplubix <tavplubix@gmail.com>

nikitamikhaylov committed:
comments + fixes for parsing

nikitamikhaylov committed:
delete stream parsing

nikitamikhaylov committed:
cmake

nikitamikhaylov committed:
better test + fix writePrefix in JSONCompact

nikitamikhaylov committed:
pvs check

nikitamikhaylov committed:
delete PrepareAndEndUpReadBuffer

nikitamikhaylov committed:
better

nikitamikhaylov committed:
disable progress bar with parallel_formatting

nikitamikhaylov committed:
get rid of std::terminate

nikitamikhaylov committed:
better

nikitamikhaylov committed:
rebase and fix tests

nikitamikhaylov committed:
fix tests

nikitamikhaylov committed:
add some logs

nikitamikhaylov committed:
thread tests

nikitamikhaylov commented:
```bash
2020-12-15 02:21:19 b"../src/IO/WriteIntText.h:27:9: runtime error: applying non-zero offset 40 to null pointer\n   
 #0 0xf1d576c in void DB::writeIntText<char8_t>(char8_t, DB::WriteBuffer&) /build/obj-x86_64-linux-gnu/../src/IO/WriteIntText.h:27:9\n    
#1 0x1a496b8d in DB::IRowOutputFormat::write(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > const&, unsigned long) /build/obj-x86_64-linux-gnu/../src/Processors/Formats/IRowOutputFormat.cpp:85:9\n    
#2 0x1a495e1f in DB::IRowOutputFormat::consume(DB::Chunk) /build/obj-x86_64-linux-gnu/../src/Processors/Formats/IRowOutputFormat.cpp:25:9\n    
#3 0x1a3387dc in DB::ParallelFormattingOutputFormat::formatterThreadFunction(unsigned long) /build/obj-x86_64-linux-gnu/../src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h:304:32\n    
#4 0xe641c05 in std::__1::function<void ()>::operator()() const /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:2473:12\n    
#5 0xe641c05 in ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) /build/obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:236:17\n    
#6 0xe646747 in void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()::operator()() const /build/obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:117:73\n    
#7 0xe646747 in decltype(std::__1::forward<void>(fp)(std::__1::forward<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&>(fp0)...)) std::__1::__invoke_constexpr<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&...) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/type_traits:3525:1\n    
#8 0xe646747 in decltype(auto) std::__1::__apply_tuple_impl<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&, std::__1::tuple<>&>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&, std::__1::__tuple_indices<std::__1::tuple<>&...>) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/tuple:1415:1\n    
#9 0xe646747 in decltype(auto) std::__1::apply<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&, std::__1::tuple<>&>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/tuple:1424:1\n    
#10 0xe646747 in ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() /build/obj-x86_64-linux-gnu/../src/Common/ThreadPool.h:177:13\n    
#11 0xe64030c in std::__1::function<void ()>::operator()() const /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:2473:12\n    
#12 0xe64030c in ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) /build/obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:236:17\n    
#13 0xe643f05 in void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()::operator()() const /build/obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:117:73\n    
#14 0xe643f05 in decltype(std::__1::forward<void>(fp)(std::__1::forward<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(fp0)...)) std::__1::__invoke<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/type_traits:3519:1\n   
 #15 0xe643f05 in void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>&, std::__1::__tuple_indices<>) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/thread:273:5\n    
#16 0xe643f05 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/thread:284:5\n   
 #17 0x7fae57a93668 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x9668)\n    #18 0x7fae579bb2b2 in clone (/lib/x86_64-linux-gnu/libc.so.6+0x1222b2)\n\nSUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/IO/WriteIntText.h:27:9 in \n"
```

Commit:
fix tsan

Commit:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into parallel-parsing-input-format

Commit:
build after merge

Commit:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into parallel-parsing-input-format

Commit:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into parallel-parsing-input-format

Commit:
Fix

Commit:
ya make update

Commit:
better

Commit:
style

Commit:
better

Commit:
style

Commit:
disable values for parallel_formatting

filimonov committed:
Attempt to prevent test flap 

That extra SELECT is executed synchronously, after that all further actions will have a more stable timing.

tavplubix commented on src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h:
```diff
@@ -0,0 +1,357 @@
+
+            auto formatter = internal_formatter_creator(out_buffer);
+
+            switch (unit.type)
```
It's not clear why reordering of processing types is not possible in current implementation, could you please add a comment?

nikitamikhaylov replied:
Added a comment. In short, there are 'two pointers' (Collector thread and thread which consumes Chunks) and first is guaranteed to be less or equal than other.

nikitamikhaylov commented on src/Common/ThreadPool.h:
```diff
@@ -196,13 +195,19 @@ class ThreadFromGlobalPool
     {
         if (joinable())
+        {
+            std::cerr << StackTrace().toString() << std::endl;
```
I want to save it, because is nothing good in doing `std::terminate` without debug info. Maybe better to use logger.

nikitamikhaylov replied:
But the log may not be flushed :(

tavplubix replied:
Maybe use `abort()` instead of `std::terminate()`? It will log stacktrace the usual way

alexey-milovidov replied:
std::terminate calls abort under the hood.

nikitamikhaylov commented on src/Interpreters/Context.cpp:
```diff
@@ -2085,7 +2086,12 @@ void Context::checkPartitionCanBeDropped(const String & database, const String &
+    return std::make_shared<InputStreamFromInputFormat>(FormatFactory::instance().getInput(name, buf, sample, *this, max_block_size));
+}
+
+BlockOutputStreamPtr Context::getOutputFormatParallelIfPossible(const String & name, WriteBuffer & buf, const Block & sample) const
```
Hmm, didn't think about it. Will fix

nikitamikhaylov commented on tests/queries/0_stateless/01514_parallel_formatting.sql:
```diff
@@ -0,0 +1,20 @@
+set output_format_parallel_formatting=1;
+create table tsv(a int, b int default 7) engine File(TSV);
+
+insert into tsv(a) select number from numbers(10000000);
```
Yes, will add test with formatting `hits` or `visits` table and calculation md5 checksum

nikitamikhaylov commented on src/Formats/FormatFactory.cpp:
```diff
@@ -214,22 +211,53 @@ BlockOutputStreamPtr FormatFactory::getOutput(const String & name,
+
+    if (!getCreators(name).output_processor_creator)
+    {
+        throw Exception("Format " + name + " is not suitable for output (with processors)", ErrorCodes::FORMAT_IS_NOT_SUITABLE_FOR_OUTPUT);
```
Only Native and Null. Both have alternatives with Processors. Maybe delete them? Better in other PR probably.

nikitamikhaylov replied:
But there is exactly the same code in master https://github.com/ClickHouse/ClickHouse/blob/006d045ba3789044667f293e5150b26be47f15cd/src/Formats/FormatFactory.cpp#L275-L277

nikitamikhaylov replied:
For now I've added separate methods for OutputStreams and OutputFormats. I was suprised, that OutputStreams are used in many places. It is a totally copy-paste, I will try to remove Streams from the codebase ASAP.

tavplubix commented:
Does parallel formatting work with HTTP interface?

nikitamikhaylov commented on src/Formats/FormatFactory.cpp:
```diff
@@ -214,22 +211,53 @@ BlockOutputStreamPtr FormatFactory::getOutput(const String & name,
+    const Settings & settings = context.getSettingsRef();
+    bool parallel_formatting = settings.output_format_parallel_formatting;
+
+    if (parallel_formatting && getCreators(name).supports_parallel_formatting && !settings.allow_experimental_live_view && !settings.output_format_json_array_of_rows)
```
Note, that parallel formatting is disabled in interactive client mode. Is this behaviour possible when using clickhouse-client from console?

tavplubix replied:
I don't know, I mean that we should rather add better protection from using parallel formatting with live views or remove this condition (if it's enough to disable parallel formatting in interactive mode)

nikitamikhaylov replied:
The problem not exactly in LiveView, but in tests for it. They are written with `uexpect` (as I understand it is an independent [framework](https://pypi.org/project/testflows.uexpect/) from TestFlows). And these tests expect Progress packets or whatever related to progress. Also, progress bar it not strictly synchronized with WriteBuffer in Client, so progress bar is mixed with data. It looks terribly, so we don't use parallel formatting in this case.

nikitamikhaylov commented on src/Formats/FormatFactory.cpp:
```diff
@@ -214,22 +211,53 @@ BlockOutputStreamPtr FormatFactory::getOutput(const String & name,
+        const auto & output_getter = getCreators(name).output_processor_creator;
 
-        /**  Materialization is needed, because formats can use the functions `IDataType`,
+        /** TODO: Materialization is needed, because formats can use the functions `IDataType`,
```
I forgot..

nikitamikhaylov replied:
It is a copy-pasted comment from another method :(

nikitamikhaylov commented:
> Does parallel formatting work with HTTP interface?

It seems like yes, but only in my local tests.

nikitamikhaylov commented on src/IO/BufferWithOwnMemory.h:
```diff
@@ -149,4 +149,35 @@ class BufferWithOwnMemory : public Base
+    {
+        const size_t prev_size = memory.size();
+        memory.resize(2 * prev_size + 1);
+        Base::set(memory.data() + prev_size, memory.size() - prev_size, 0);
```
Really, I think we can calculate prev_size as `Base::position() - memory.data()` and it will help

Commit:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into parallel-parsing-input-format

Commit:
add tests

Commit:
fix memory and add comment

Commit:
better formatfactory

Commit:
arcadia fix + live view fix + cleanup

nikitamikhaylov commented on src/Processors/Formats/Impl/ParallelParsingInputFormat.cpp:
```diff
@@ -0,0 +1,224 @@
+        if (current_ticket_number == 0)
+            prepareReadBuffer(read_buffer);
+
+        InputFormatPtr input_format = internal_parser_creator(read_buffer);
```
But it's a little different, I think. We still need to recreate the ReadBuffer and put it in InputFormat, because otherwise freeing the buffer or resetting its state is very difficult due to the inconvenient API.

nikitamikhaylov commented on src/Processors/Formats/Impl/IReadBufferPrepareAndEndUp.h:
```diff
@@ -0,0 +1,53 @@
+namespace DB
+{
+
+class IReadBufferPrepareAndEndUp
```
I've got rid from it.

nikitamikhaylov commented on src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp:
```diff
@@ -245,16 +245,16 @@ bool JSONEachRowRowInputFormat::readRow(MutableColumns & columns, RowReadExtensi
     /// then seeking to next ;, or \n would trigger reading of an extra row at the end.
 
     /// Semicolon is added for convenience as it could be used at end of INSERT query.
-    bool is_first_row = getCurrentUnitNumber() == 0 && getTotalRows() == 1;
```
There was PrepareAndEndUpReadBuffer class, but it did't solve the problem. I removed it and forgot to return all back.

Commit:
abort() instead of std::terminate() + cleanup

Commit:
better

Commit:
cleanup

Commit:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into parallel-parsing-input-format

Commit:
build fix

nikitamikhaylov committed:
remove kafka flaky fix

Commit:
shellchek

Commit:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into parallel-parsing-input-format

Commit:
Merge branch 'parallel-parsing-input-format' of github.com:nikitamikhaylov/ClickHouse into parallel-parsing-input-format

Commit:
shellcheck

Commit:
fix build

nikitamikhaylov committed:
Merge branch 'master' into parallel-parsing-input-format

nikitamikhaylov commented:
Performance test is broken in master. 01502_bar_overflow (ubsan) is broken is master too. 01133_max_result_rows (memory) is a bit flaky. AST Fuzzer failed without any stacktrace.

sevirov commented:
Internal documentation ticket: DOCSUP-13574

alexey-milovidov commented:
https://github.com/ClickHouse/ClickHouse/issues/30873
